### ISSUE TITLE
erroneous GeoJSON type-guessing

### ISSUE THREAD

**Issue by kgjenkins** [qgis/QGIS#33650]
**Describe the bug**

When opening a GeoJSON file containing a property that has string values that look somewhat like dates, QGIS erroneously interprets them as QDate types.

For example, using this GeoJSON:
```
{
  "type": "FeatureCollection",
  "name": "test",
  "features": [
    {
      "type": "Feature",
      "properties": {
        "identifier": "3000-02-31",
        "name":"a"
      },
      "geometry": { "type": "Point", "coordinates": [ 1, 1 ] }
    },
    {
      "type": "Feature",
      "properties": {
        "identifier": "9000-01-01",
        "name":"b"
      },
      "geometry": { "type": "Point", "coordinates": [ 2, 1 ] }
    }
  ]
}
```

The layer properties > fields shows:
![image](https://user-images.githubusercontent.com/3355358/71923749-b663cc00-315b-11ea-9dbb-9c94880545fa.png)

And the attribute table looks like this:
![image](https://user-images.githubusercontent.com/3355358/71923860-e612d400-315b-11ea-839b-79b66ede8b82.png)

If any changes are made in QGIS, then the values in the modified file get mangled:
* `3000-02-31` becomes `3000\/02\/31`
* `9000-01-01` becomes `9000\/01\/01`

**How to Reproduce**

1. Load the test.geojson file in [test.zip](https://github.com/qgis/QGIS/files/4032180/test.zip)

2. Right-click > Layer properties > Fields, and note the QData type

3. Open the attribute table, and note the erroneous "null"

4. Turn on editing, move one of the points, and save.  Then open the modified test.geojson file in a text editor and note the mangled "identifier" values.

**QGIS and OS versions**

QGIS version 3.10.1-A Coruña
QGIS code revision ef24c526da
Compiled against Qt 5.11.2
Running against Qt 5.11.2
Compiled against GDAL/OGR 3.0.2
Running against GDAL/OGR 3.0.2
Compiled against GEOS 3.8.0-CAPI-1.13.1
Running against GEOS 3.8.0-CAPI-1.13.1 
Compiled against SQLite 3.29.0
Running against SQLite 3.29.0
PostgreSQL Client Version 11.5
SpatiaLite Version 4.3.0
QWT Version 6.1.3
QScintilla2 Version 2.10.8
Compiled against PROJ 6.2.1
Running against PROJ Rel. 6.2.1, November 1st, 2019
OS Version Windows 10 (10.0)

**Additional context**

I originally came across this behavior with actual date strings where the formatting was getting mangled from "yyyy-mm-dd" to "yyyy\/mm\/dd", then experimented a bit to discover that even non-valid dates like "3000-02-31" and "9000-01-01" get interpreted as dates.  It seems that it would be safer to just treat all GeoJSON strings as strings, or at least have an option for such behavior; otherwise it doesn't seem possible to edit a GeoJSON file safely.

**Comment by nyalldawson** [comment]
This is coming straight from gdal -- it's reporting those fields as date fields

**Comment by kgjenkins** [comment]
Thanks, I now see that ogr2ogr does the same thing.  I'll take this to the gdal-dev list.

**Comment by kgjenkins** [comment]
GDAL 3.0.3 added a `DATE_AS_STRING` read option on the GeoJSON driver that can be set to `YES` to disable autodetection of date/time/datetime.

Would it be possible/advisable for QGIS to use this option by default?  That would prevent date-like strings from getting silently mangled when editing in QGIS (as in the example above).

If implemented, one possible disadvantage would be that dates in GeoJSON strings would not automatically be converted to dates, for example when saving to other formats.  But dates are not actually a datatype in GeoJSON, and that is what Refactor Fields is for, right?

**Comment by kgjenkins** [comment]
@nyalldawson Could you please re-open this issue?  Thanks, as I cannot.

**Comment by rduivenvoorde** [comment]
@kgjenkins mmm, as somebody who uses geojson in measurement-environments (SensorThingsApi) in which the data is full of (iso)datetime fields, which are now automagically recognized as dates so I can use them in the temporal control etc without having to do anything..., I'm not very eager to have to change data every time :-)

But I agree that it would be nice to be able to have an option to set DATE_AS_STRING to YES *somewhere*. 

Note that when you EXPORT something to geojson, QGIS offers you to set some (driver dependent) Layer Options:

![Screenshot-20200608121040-808x972](https://user-images.githubusercontent.com/731673/84019135-2e3f5800-a981-11ea-9648-cc21c41c853b.png)

So IF we would add an way to set (gdal) Layer Options somewhere, question is where would it best fit:

- as some global (hidden/advanced) setting (looks easier to implement? Could be the first step, so your problem is fixed! ). In this way you could via Options/Advanced setting set the default DATE_AS_STRING from default false to your preferred true. This would then work 'globally' for all geojson's... I think... (/me wondering if QGIS already knows the format before giving the data to gdal....)

- some 'Layer Options' block in the Vector dialog just below the Encoding item, behaving like the one in the export function, so depending on your Format, is filled with different options

![Screenshot-20200608120958-754x278](https://user-images.githubusercontent.com/731673/84019572-b887bc00-a981-11ea-854c-fad5a070283e.png)

This last option is (my guess) much harder to implement, and with only one use-case at this moment does probably not make this a good case?
Also the drag-drop loading behaviour of geojson then will have another behaviour as the open from dialog, but that is maybe not an argument because that is already so with other formats too...

**Comment by rduivenvoorde** [comment]
Mmm, I see there are already soem gdal (raster) options in Advanced Settings:

![Screenshot-20200608122525-931x769](https://user-images.githubusercontent.com/731673/84020569-50d27080-a983-11ea-854c-81090d77a2c2.png)

**Comment by kgjenkins** [comment]
I've noticed that in the latest version of QGIS 3.13.0-Master (59ce5c8d65), editing a GeoJSON file no longer converts dates into "yyyy/mm/dd" format, but instead uses proper ISO 8601 format "yyyy-mm-dd", which was my main concern.

Not sure if this is due to changes in QGIS or GDAL.  (I searched both projects, but couldn't find the relevant commit.)  In any case, I'm satisfied with the current behaviour and am closing this issue.

(The handling of non-date-but-date-looking strings ("3000-02-31") is still an issue, but was an invented, unlikely edge case.  Someone else can open another issue if it really poses a problem.)
